src/hotspot/share/utilities/hashtable.hpp

```diff
@@ -242,15 +242,35 @@ class KVHashtable : public BasicHashtable<F> {
     return (KVHashtableEntry*)BasicHashtable<F>::bucket(i);
   }
 
+  // The following method is not MT-safe and must be done under lock.
+  KVHashtableEntry** bucket_addr(int i) {
+    return (KVHashtableEntry**)BasicHashtable<F>::bucket_addr(i);
+  }
+
   KVHashtableEntry* new_entry(unsigned int hashValue, K key, V value) {
     KVHashtableEntry* entry = (KVHashtableEntry*)BasicHashtable<F>::new_entry(hashValue);
     entry->_key   = key;
     entry->_value = value;
     return entry;
   }
 
+  void free_entry(KVHashtableEntry* entry) {
+    BasicHashtable<F>::free_entry(entry);
+  }
+
 public:
   KVHashtable(int table_size) : BasicHashtable<F>(table_size, sizeof(KVHashtableEntry)) {}
+  ~KVHashtable() {
+    KVHashtableEntry* probe = NULL;
+    for (int index = 0; index < table_size(); index++) {
+      for (KVHashtableEntry** p = bucket_addr(index); *p != NULL; ) {
+        probe = *p;
+        *p = probe->next();
+        free_entry(probe);
```

Comment:
I hope unlink_entry isn't wrong because it's used for all the BasicHashtables.  It does orphan the further entries, but this entry has already been removed from the list. So setting the next isn't really necessary here, but decrementing the number of elements is for printing and logging.